python/tank/util/storage_roots.py

```diff
@@ -0,0 +1,615 @@
+        unmapped_root_names = []
+
+        # query all local storages from SG so we can store a lookup of roots
+        # defined here to SG storages
+
+        # fields required for SG local storage queries
+        local_storage_fields = ["code", "id"]
+        local_storage_fields.extend(ShotgunPath.SHOTGUN_PATH_FIELDS)
+
+        # create the SG connection and query
+        log.debug("Querying SG local storages...")
+        sg_storages = sg_connection.find(
+            "LocalStorage",
+            [],
+            local_storage_fields
+        )
+        log.debug("Query returned %s storages." % (len(sg_storages, )))
+
+        # create lookups of storages by name and id for convenience. we'll check
+        # against each root's shotgun_storage_id first, falling back to the
+        # name if the id mapping field is not defined.
+        sg_storages_by_id = {}
+        sg_storages_by_name = {}
+        for sg_storage in sg_storages:
+            id = sg_storage["id"]
+            name = sg_storage["code"]
+            sg_storages_by_id[id] = sg_storage
+            sg_storages_by_name[name] = sg_storage
+
+        for root_name, root_info in self:
+
+            # see if the shotgun storage id is specified explicitly in the
+            # roots.yml file.
+            root_storage_id = root_info.get("shotgun_storage_id")
+            if root_storage_id and root_storage_id in sg_storages_by_id:
+                # found a match. store it in the lookup
+                sg_storage = sg_storages_by_id[root_storage_id]
+                log.debug(
+                    "Storage root %s explicitly associated with SG local "
+                    "storage id %s (%s)" %
+                    (root_name, root_storage_id, sg_storage)
+                )
+                local_storage_lookup[root_name] = sg_storage
+                continue
+
+            # if we're here, no storage is specified explicitly. fall back to
+            # the storage name.
+            if root_name in sg_storages_by_name:
+                # found a match. store it in the lookup
+                sg_storage = sg_storages_by_name[root_name]
+                log.debug(
+                    "Storage root %s matches SG local storage with same name "
+                    "(%s)" % (root_name, sg_storage)
+                )
+                local_storage_lookup[root_name] = sg_storage
+                continue
+
+            # if we're here, then we could not map the storage root to a local
+            # storage in SG
+            log.warning(
+                "Storage root %s could not be mapped to a SG local storage" %
+                (root_name,)
+            )
+            unmapped_root_names.append(root_name)
+
+        # return a tuple of the processed info
+        return local_storage_lookup, unmapped_root_names
+
+    def populate_defaults(self):
+        """
+        This method ensures all sys.platforms are represented in all defined
```

Comment:
Unless it's a mess to rewrite the code from the setup_project command using the new class because not enough stuff is exposed (and that shouldn't be exposed) then I'd suggest it stays in setup_project. It definitely feels like a method intended for a very specific feature and not a general use like the class has been so far.